File: max_counters/max_counters.js
// https://app.codility.com/demo/results/trainingG6BPDV-4YE/

function solution(N, A) {
  let counters = new Array(N).fill(0);
  let max = 0;
  let reset = 0;

  for (let i = 0; i < A.length; i++) {
    if (A[i] > 0 && A[i] <= N) {
      let ref = counters[A[i] - 1];
      ref = ref > reset ? ref + 1 : reset + 1;
      counters[A[i] - 1] = ref;
      max = ref > max ? ref : max;
    } else {
      reset = max;
    }
  }

  for (let i = 0; i < N; i++) {
    if (counters[i] < reset) {
      counters[i] = reset;
    }
  }

  return counters;
}

File: max_counters/max_counters.py
# https://app.codility.com/demo/results/trainingW8Y4HV-GDD/


def solution(N, A):
    containers = [0] * N
    max_counter = 0
    current_max = 0

    for value in A:
        if (value > N):
            max_counter = current_max
        else:
            if (containers[value - 1] < max_counter):
                containers[value - 1] = max_counter

            containers[value - 1] = containers[value - 1] + 1

            if (current_max < containers[value - 1]):
                current_max = containers[value - 1]

    for i in range(len(containers)):
        if (containers[i] < max_counter):
            containers[i] = max_counter

    return containers
